src/PostgREST/Main.hs

```diff
@@ -105,7 +99,8 @@ main = do
   -- print $ findRelation (fakeRels ++ allRels) "test" "pg_source" "clients"
 
   runSettings appSettings $ middle $ \ req respond -> do
+    time <- getPOSIXTime
     body <- strictRequestBody req
     resOrError <- liftIO $ H.session pool $ H.tx txSettings $
-      runWithClaims conf (app dbstructure conf body) req
+      runWithClaims conf time (app dbstructure conf body) req
```

Comment:
I think it needs to go inside the middleware itself, like the code snippet below. There was one dirty little trick though, the `unsafePerformIO`. You see the hasql library intentionally prohibits us from calling IO actions inside their `Tx` monad because

> Running IO in Tx is prohibited. The motivation is identical to STM: the Tx block may get executed multiple times if any transaction conflicts arise. This will result in your effectful IO code being executed an unpredictable amount of times as well, which, chances are, is not what you want.

In our case the action is just to ask the time, and I don't care if that action happens to get called multiple times. The unsafe perform is generally a Haskell no-no but I don't know what else to do. We need to get the time of the request itself and not the time of server start.

Testing this is also hard. We would have to mock the computer clock inside the tests. I don't know if this is possible in Haskell.

Here is the code change to move the time checking inside the middleware to be executed as each request happens.

```diff
diff --git a/src/PostgREST/Main.hs b/src/PostgREST/Main.hs
index f51a936..1a49e98 100644
--- a/src/PostgREST/Main.hs
+++ b/src/PostgREST/Main.hs
@@ -2,7 +2,6 @@ module Main where
 
 
 import           PostgREST.App
--- import PostgREST.QueryBuilder
 import           PostgREST.Config                     (AppConfig (..),
                                                        minimumPgVersion,
                                                        prettyVersion,
@@ -24,7 +23,6 @@ import qualified Hasql.Postgres                       as P
 import           Network.Wai
 import           Network.Wai.Handler.Warp             hiding (Connection)
 import           Network.Wai.Middleware.RequestLogger (logStdout)
-import           Data.Time.Clock.POSIX                (getPOSIXTime)
 import           System.IO                            (BufferMode (..),
                                                        hSetBuffering, stderr,
                                                        stdin, stdout)
@@ -99,8 +97,7 @@ main = do
   -- print $ findRelation (fakeRels ++ allRels) "test" "pg_source" "clients"
 
   runSettings appSettings $ middle $ \ req respond -> do
-    time <- getPOSIXTime
     body <- strictRequestBody req
     resOrError <- liftIO $ H.session pool $ H.tx txSettings $
-      runWithClaims conf time (app dbstructure conf body) req
+      runWithClaims conf (app dbstructure conf body) req
     either (respond . errResponse) respond resOrError
diff --git a/src/PostgREST/Middleware.hs b/src/PostgREST/Middleware.hs
index 9d4ea5f..aaaf748 100644
--- a/src/PostgREST/Middleware.hs
+++ b/src/PostgREST/Middleware.hs
@@ -7,7 +7,7 @@ import           Data.Maybe                    (fromMaybe, isNothing)
 import           Data.Monoid
 import           Data.Text
 import           Data.String.Conversions       (cs)
-import           Data.Time.Clock               (NominalDiffTime)
+import           Data.Time.Clock.POSIX         (getPOSIXTime)
 import qualified Hasql                         as H
 import qualified Hasql.Postgres                as P
 
@@ -27,20 +27,23 @@ import           PostgREST.App                 (contentTypeForAccept)
 import           PostgREST.Auth                (setRole, jwtClaims, claimsToSQL)
 import           PostgREST.Config              (AppConfig (..), corsPolicy)
 
+import           System.IO.Unsafe              (unsafePerformIO)
+
 import           Prelude hiding(concat)
 
 import qualified Data.Vector             as V
 import qualified Hasql.Backend           as B
 import qualified Data.Map.Lazy           as M
 
-runWithClaims :: forall s. AppConfig -> NominalDiffTime ->
+runWithClaims :: forall s. AppConfig ->
                  (Request -> H.Tx P.Postgres s Response) ->
                  Request -> H.Tx P.Postgres s Response
-runWithClaims conf time app req = do
+runWithClaims conf app req = do
     _ <- H.unitEx $ stmt setAnon
+    let t = unsafePerformIO getPOSIXTime
     case split (== ' ') (cs auth) of
       ("Bearer" : tokenStr : _) ->
-        case jwtClaims jwtSecret tokenStr time of
+        case jwtClaims jwtSecret tokenStr t of
           Just claims ->
             if M.member "role" claims
             then do
```